DataDog/extendeddaemonset
# Add telemetry to detect stuck rolling updates

### What does this PR do?

Add a new metric `ers_rolling_update_stuck`: `1` if the number of unavailable pods is higher than maxUnavailable, `0` otherwise.

### Motivation

To be alerted if an ERS rolling update get stuck because of the number of unavailable pods in the cluster.

### Describe your test plan

- Deploy the EDS
- Apply a CR with a low maxUnavailable value (min supported value is 1)
```
apiVersion: datadoghq.com/v1alpha1
kind: ExtendedDaemonSet
metadata:
  name: pause-containers
spec:
  strategy:
    rollingUpdate:
      maxUnavailable: 1
  template:
    spec:
      containers:
      - name: daemon
        image: k8s.gcr.io/pause:3.0
```
- After the deployment is done, trigger a rolling update that will cause the new pods to be available,. To do so, a broken readiness probe can be added. This will make the EDS consider the new pod unavailable and should throttle/stuck the rolling update
```
apiVersion: datadoghq.com/v1alpha1
kind: ExtendedDaemonSet
metadata:
  name: pause-containers
spec:
  strategy:
    rollingUpdate:
      maxUnavailable: 1
  template:
    spec:
      containers:
      - name: daemon
        image: k8s.gcr.io/pause:3.0
        readinessProbe:
          tcpSocket:
            port: 8080
```
- Make sure the metric shows that the new ers is stuck
```
$ curl localhost:8080/metrics | grep ers_rolling_update_stuck
# HELP ers_rolling_update_stuck 1 if the number of unavailable pods is higher than maxUnavailable, 0 otherwise
# TYPE ers_rolling_update_stuck gauge
ers_rolling_update_stuck{name="pause-containers-67zbv",namespace="default"} 0
ers_rolling_update_stuck{name="pause-containers-vpkxk",namespace="default"} 1
```
